File: src/models/monte_carlo.py
import numpy as np
from typing import Tuple
from joblib import Parallel, delayed
import time

class MonteCarlo:
    """
    Monte Carlo simulation for European option pricing.
    
    Uses geometric Brownian motion to simulate stock price paths
    and calculates option payoffs at maturity.
    """
    
    def __init__(self, S: float, K: float, T: float, r: float, sigma: float, 
                 n_simulations: int = 100000, n_jobs: int = -1):
        """
        Initialize Monte Carlo pricing model.
        
        Parameters:
        S: Current stock price
        K: Strike price
        T: Time to maturity (years)
        r: Risk-free rate
        sigma: Volatility
        n_simulations: Number of price paths to simulate
        n_jobs: Number of parallel jobs (-1 = all CPUs)
        """
        self.S = S
        self.K = K
        self.T = T
        self.r = r
        self.sigma = sigma
        self.n_simulations = n_simulations
        self.n_jobs = n_jobs
        self.simulated_prices = None
        
    def _simulate_batch(self, batch_size: int, seed: int) -> np.ndarray:
        """Simulate a batch of terminal stock prices."""
        np.random.seed(seed)
        
        # Generate random standard normal variables
        Z = np.random.standard_normal(batch_size)
        
        # Calculate terminal stock price using GBM formula
        # S_T = S_0 * exp((r - 0.5*sigma^2)*T + sigma*sqrt(T)*Z)
        drift = (self.r - 0.5 * self.sigma**2) * self.T
        diffusion = self.sigma * np.sqrt(self.T) * Z
        
        S_T = self.S * np.exp(drift + diffusion)
        return S_T
    
    def simulate(self, verbose: bool = True) -> None:
        """Run Monte Carlo simulation in parallel."""
        if verbose:
            print(f"\nRunning Monte Carlo with {self.n_simulations:,} simulations...")
            print(f"Using parallel execution with n_jobs={self.n_jobs}")
        
        start_time = time.time()
        
        # Determine batch size for parallel execution
        n_cores = self.n_jobs if self.n_jobs > 0 else -1
        
        if n_cores == -1:
            import os
            n_cores = os.cpu_count()
        
        batch_size = self.n_simulations // n_cores
        
        # Run simulations in parallel
        results = Parallel(n_jobs=self.n_jobs)(
            delayed(self._simulate_batch)(batch_size, seed=i) 
            for i in range(n_cores)
        )
        
        # Combine results
        self.simulated_prices = np.concatenate(results)
        
        # Handle remainder if n_simulations doesn't divide evenly
        remainder = self.n_simulations - len(self.simulated_prices)
        if remainder > 0:
            extra_prices = self._simulate_batch(remainder, seed=n_cores)
            self.simulated_prices = np.concatenate([self.simulated_prices, extra_prices])
        
        elapsed = time.time() - start_time
        
        if verbose:
            print(f"Simulation completed in {elapsed:.3f} seconds")
            print(f"Simulations per second: {self.n_simulations/elapsed:,.0f}")
    
    def call_price(self) -> float:
        """Calculate call option price from simulations."""
        if self.simulated_prices is None:
            raise ValueError("Must run simulate() first")
        
        # Call payoff: max(S_T - K, 0)
        payoffs = np.maximum(self.simulated_prices - self.K, 0)
        
        # Discount to present value
        return np.exp(-self.r * self.T) * np.mean(payoffs)
    
    def put_price(self) -> float:
        """Calculate put option price from simulations."""
        if self.simulated_prices is None:
            raise ValueError("Must run simulate() first")
        
        # Put payoff: max(K - S_T, 0)
        payoffs = np.maximum(self.K - self.simulated_prices, 0)
        
        # Discount to present value
        return np.exp(-self.r * self.T) * np.mean(payoffs)
    
    def confidence_interval(self, option_type: str = 'call', 
                           confidence: float = 0.95) -> Tuple[float, float]:
        """
        Calculate confidence interval for option price.
        
        Parameters:
        option_type: 'call' or 'put'
        confidence: Confidence level (e.g., 0.95 for 95%)
        """
        if self.simulated_prices is None:
            raise ValueError("Must run simulate() first")
        
        # Calculate payoffs
        if option_type.lower() == 'call':
            payoffs = np.maximum(self.simulated_prices - self.K, 0)
        else:
            payoffs = np.maximum(self.K - self.simulated_prices, 0)
        
        # Discount payoffs
        discounted_payoffs = np.exp(-self.r * self.T) * payoffs
        
        # Calculate standard error
        mean_price = np.mean(discounted_payoffs)
        std_error = np.std(discounted_payoffs) / np.sqrt(self.n_simulations)
        
        # Confidence interval (using normal approximation)
        from scipy.stats import norm
        z_score = norm.ppf((1 + confidence) / 2)
        margin = z_score * std_error
        
        return (mean_price - margin, mean_price + margin)
    
    def summary(self) -> None:
        """Print Monte Carlo pricing summary."""
        if self.simulated_prices is None:
            print("No simulation results. Run simulate() first.")
            return
        
        print("=" * 60)
        print("MONTE CARLO OPTION PRICING")
        print("=" * 60)
        print(f"\nInput Parameters:")
        print(f"  Stock Price (S):      ${self.S:.2f}")
        print(f"  Strike Price (K):     ${self.K:.2f}")
        print(f"  Time to Maturity (T): {self.T:.4f} years ({self.T*365:.0f} days)")
        print(f"  Risk-free Rate (r):   {self.r*100:.2f}%")
        print(f"  Volatility (σ):       {self.sigma*100:.2f}%")
        print(f"  Number of Sims:       {self.n_simulations:,}")
        
        call = self.call_price()
        put = self.put_price()
        
        call_ci = self.confidence_interval('call')
        put_ci = self.confidence_interval('put')
        
        print(f"\nSimulation Statistics:")
        print(f"  Mean Terminal Price:  ${np.mean(self.simulated_prices):.2f}")
        print(f"  Std Terminal Price:   ${np.std(self.simulated_prices):.2f}")
        print(f"  Min Terminal Price:   ${np.min(self.simulated_prices):.2f}")
        print(f"  Max Terminal Price:   ${np.max(self.simulated_prices):.2f}")
        
        print(f"\nOption Prices:")
        print(f"  Call Option: ${call:.4f}")
        print(f"    95% CI: (${call_ci[0]:.4f}, ${call_ci[1]:.4f})")
        print(f"  Put Option:  ${put:.4f}")
        print(f"    95% CI: (${put_ci[0]:.4f}, ${put_ci[1]:.4f})")
        
        print("=" * 60)


# Test and compare with Black-Scholes
if __name__ == "__main__":
    from black_scholes import BlackScholes
    
    print("\n" + "="*60)
    print("MONTE CARLO vs BLACK-SCHOLES COMPARISON")
    print("="*60)
    
    # Test parameters
    S, K, T, r, sigma = 100, 100, 1.0, 0.05, 0.20
    
    # Black-Scholes (analytical)
    print("\n1. BLACK-SCHOLES (Analytical Solution)")
    bs = BlackScholes(S, K, T, r, sigma)
    bs_call = bs.call_price()
    bs_put = bs.put_price()
    print(f"   Call: ${bs_call:.4f}")
    print(f"   Put:  ${bs_put:.4f}")
    
    # Monte Carlo with different simulation counts
    sim_counts = [10000, 100000, 1000000]
    
    for n_sims in sim_counts:
        print(f"\n2. MONTE CARLO ({n_sims:,} simulations)")
        mc = MonteCarlo(S, K, T, r, sigma, n_simulations=n_sims)
        mc.simulate(verbose=True)
        
        mc_call = mc.call_price()
        mc_put = mc.put_price()
        
        print(f"   Call: ${mc_call:.4f} (Error: ${abs(mc_call - bs_call):.4f} or {abs(mc_call - bs_call)/bs_call*100:.2f}%)")
        print(f"   Put:  ${mc_put:.4f} (Error: ${abs(mc_put - bs_put):.4f} or {abs(mc_put - bs_put)/bs_put*100:.2f}%)")
        
        call_ci = mc.confidence_interval('call')
        print(f"   Call 95% CI: (${call_ci[0]:.4f}, ${call_ci[1]:.4f})")
    
    print("\n" + "="*60)

File: java/src/main/java/com/options/models/MonteCarlo.java
package com.options.models;

import org.apache.commons.math3.distribution.NormalDistribution;
import org.apache.commons.math3.random.MersenneTwister;
import java.util.concurrent.RecursiveTask;
import java.util.concurrent.ForkJoinPool;

/**
 * Monte Carlo simulation for European option pricing.
 * 
 * Uses geometric Brownian motion to simulate stock price paths
 * and calculates option payoffs at maturity with parallel processing.
 */
public class MonteCarlo {
    
    private final double S;              // Current stock price
    private final double K;              // Strike price
    private final double T;              // Time to maturity (years)
    private final double r;              // Risk-free rate
    private final double sigma;          // Volatility
    private final int nSimulations;      // Number of price paths to simulate
    
    private double[] simulatedPrices;    // Terminal stock prices
    private final NormalDistribution normal;
    private final MersenneTwister random;
    
    /**
     * Initialize Monte Carlo pricing model.
     * 
     * @param S Current stock price
     * @param K Strike price
     * @param T Time to maturity (years)
     * @param r Risk-free rate
     * @param sigma Volatility
     * @param nSimulations Number of price paths to simulate
     */
    public MonteCarlo(double S, double K, double T, double r, double sigma, int nSimulations) {
        this.S = S;
        this.K = K;
        this.T = T;
        this.r = r;
        this.sigma = sigma;
        this.nSimulations = nSimulations;
        this.normal = new NormalDistribution();
        this.random = new MersenneTwister();
    }
    
    /**
     * Run Monte Carlo simulation with parallel processing.
     * 
     * @param verbose Whether to print progress information
     */
    public void simulate(boolean verbose) {
        if (verbose) {
            System.out.printf("%nRunning Monte Carlo with %,d simulations...%n", nSimulations);
            System.out.printf("Using parallel execution with ForkJoinPool%n");
        }
        
        long startTime = System.currentTimeMillis();
        
        // Use ForkJoinPool for parallel processing
        SimulationTask task = new SimulationTask(0, nSimulations, random.nextInt());
        ForkJoinPool.commonPool().invoke(task);
        simulatedPrices = task.getResults();
        
        long elapsed = System.currentTimeMillis() - startTime;
        
        if (verbose) {
            System.out.printf("Simulation completed in %.3f seconds%n", elapsed / 1000.0);
            System.out.printf("Simulations per second: %,.0f%n", nSimulations / (elapsed / 1000.0));
        }
    }
    
    /**
     * Calculate call option price from simulations.
     * 
     * @return Call option price
     */
    public double callPrice() {
        if (simulatedPrices == null) {
            throw new IllegalStateException("Must run simulate() first");
        }
        
        double totalPayoff = 0;
        for (double price : simulatedPrices) {
            // Call payoff: max(S_T - K, 0)
            totalPayoff += Math.max(price - K, 0);
        }
        
        // Discount to present value
        return Math.exp(-r * T) * (totalPayoff / nSimulations);
    }
    
    /**
     * Calculate put option price from simulations.
     * 
     * @return Put option price
     */
    public double putPrice() {
        if (simulatedPrices == null) {
            throw new IllegalStateException("Must run simulate() first");
        }
        
        double totalPayoff = 0;
        for (double price : simulatedPrices) {
            // Put payoff: max(K - S_T, 0)
            totalPayoff += Math.max(K - price, 0);
        }
        
        // Discount to present value
        return Math.exp(-r * T) * (totalPayoff / nSimulations);
    }
    
    /**
     * Calculate confidence interval for option price.
     * 
     * @param optionType 'call' or 'put'
     * @param confidence Confidence level (e.g., 0.95 for 95%)
     * @return Array with [lower_bound, upper_bound]
     */
    public double[] confidenceInterval(String optionType, double confidence) {
        if (simulatedPrices == null) {
            throw new IllegalStateException("Must run simulate() first");
        }
        
        // Calculate payoffs
        double[] payoffs = new double[nSimulations];
        for (int i = 0; i < nSimulations; i++) {
            if ("call".equalsIgnoreCase(optionType)) {
                payoffs[i] = Math.max(simulatedPrices[i] - K, 0);
            } else {
                payoffs[i] = Math.max(K - simulatedPrices[i], 0);
            }
        }
        
        // Discount payoffs
        double[] discountedPayoffs = new double[nSimulations];
        double discountFactor = Math.exp(-r * T);
        for (int i = 0; i < nSimulations; i++) {
            discountedPayoffs[i] = payoffs[i] * discountFactor;
        }
        
        // Calculate statistics
        double mean = 0;
        for (double payoff : discountedPayoffs) {
            mean += payoff;
        }
        mean /= nSimulations;
        
        double variance = 0;
        for (double payoff : discountedPayoffs) {
            variance += Math.pow(payoff - mean, 2);
        }
        variance /= (nSimulations - 1);
        double stdError = Math.sqrt(variance / nSimulations);
        
        // Confidence interval (using normal approximation)
        double zScore = normal.inverseCumulativeProbability((1 + confidence) / 2);
        double margin = zScore * stdError;
        
        return new double[]{mean - margin, mean + margin};
    }
    
    /**
     * Print Monte Carlo pricing summary.
     */
    public void printSummary() {
        if (simulatedPrices == null) {
            System.out.println("No simulation results. Run simulate() first.");
            return;
        }
        
        System.out.println("==========================================================");
        System.out.println("               MONTE CARLO OPTION PRICING             ");
        System.out.println("==========================================================");
        
        System.out.printf("%nInput Parameters:%n");
        System.out.printf("  Stock Price (S):      $%.2f%n", S);
        System.out.printf("  Strike Price (K):     $%.2f%n", K);
        System.out.printf("  Time to Maturity (T): %.4f years (%.0f days)%n", T, T * 365);
        System.out.printf("  Risk-free Rate (r):   %.2f%%%n", r * 100);
        System.out.printf("  Volatility (σ):       %.2f%%%n", sigma * 100);
        System.out.printf("  Number of Sims:       %,d%n", nSimulations);
        
        double call = callPrice();
        double put = putPrice();
        
        double[] callCI = confidenceInterval("call", 0.95);
        double[] putCI = confidenceInterval("put", 0.95);
        
        // Calculate terminal price statistics
        double meanTerminal = 0;
        double minTerminal = Double.MAX_VALUE;
        double maxTerminal = Double.MIN_VALUE;
        double variance = 0;
        
        for (double price : simulatedPrices) {
            meanTerminal += price;
            if (price < minTerminal) minTerminal = price;
            if (price > maxTerminal) maxTerminal = price;
        }
        meanTerminal /= nSimulations;
        
        for (double price : simulatedPrices) {
            variance += Math.pow(price - meanTerminal, 2);
        }
        variance /= (nSimulations - 1);
        double stdTerminal = Math.sqrt(variance);
        
        System.out.printf("%nSimulation Statistics:%n");
        System.out.printf("  Mean Terminal Price:  $%.2f%n", meanTerminal);
        System.out.printf("  Std Terminal Price:   $%.2f%n", stdTerminal);
        System.out.printf("  Min Terminal Price:   $%.2f%n", minTerminal);
        System.out.printf("  Max Terminal Price:   $%.2f%n", maxTerminal);
        
        System.out.printf("%nOption Prices:%n");
        System.out.printf("  Call Option: $%.4f%n", call);
        System.out.printf("    95%% CI: ($%.4f, $%.4f)%n", callCI[0], callCI[1]);
        System.out.printf("  Put Option:  $%.4f%n", put);
        System.out.printf("    95%% CI: ($%.4f, $%.4f)%n", putCI[0], putCI[1]);
        
        System.out.println("==========================================================");
    }
    
    // Getters
    public double getS() { return S; }
    public double getK() { return K; }
    public double getT() { return T; }
    public double getR() { return r; }
    public double getSigma() { return sigma; }
    public int getNSimulations() { return nSimulations; }
    public double[] getSimulatedPrices() { return simulatedPrices; }
    
    /**
     * Recursive task for parallel Monte Carlo simulation.
     */
    private class SimulationTask extends RecursiveTask<double[]> {
        private static final int THRESHOLD = 10000; // Minimum size for parallel processing
        private final int start;
        private final int end;
        private final int seed;
        
        SimulationTask(int start, int end, int seed) {
            this.start = start;
            this.end = end;
            this.seed = seed;
        }
        
        @Override
        protected double[] compute() {
            int length = end - start;
            
            if (length <= THRESHOLD) {
                // Compute directly
                MersenneTwister localRandom = new MersenneTwister(seed);
                double[] results = new double[length];
                
                double drift = (r - 0.5 * sigma * sigma) * T;
                double diffusion = sigma * Math.sqrt(T);
                
                for (int i = 0; i < length; i++) {
                    double Z = localRandom.nextGaussian();
                    double price = S * Math.exp(drift + diffusion * Z);
                    results[i] = price;
                }
                
                return results;
            } else {
                // Split and compute in parallel
                int mid = start + length / 2;
                
                SimulationTask leftTask = new SimulationTask(start, mid, seed);
                SimulationTask rightTask = new SimulationTask(mid, end, seed + 1);
                
                leftTask.fork();
                double[] rightResult = rightTask.compute();
                double[] leftResult = leftTask.join();
                
                // Combine results
                double[] combined = new double[length];
                System.arraycopy(leftResult, 0, combined, 0, leftResult.length);
                System.arraycopy(rightResult, 0, combined, leftResult.length, rightResult.length);
                
                return combined;
            }
        }
        
        public double[] getResults() {
            return compute();
        }
    }
}